Markdown/Markdown.sublime-syntax

```diff
@@ -580,61 +580,77 @@ contexts:
   escape:
     - match: '{{escape}}'
       scope: constant.character.escape.markdown
-  atx-heading-terminator:
-    - match: '[ ]*(#*)[ ]*($\n?)' # \n is optional so ## is matched as end punctuation in new document (at eof)
-      captures:
-        1: punctuation.definition.heading.end.markdown
-        2: meta.whitespace.newline.markdown
-      pop: true
+
   atx-heading:
-    - match: '(#)(?!#)\s*(?=\S)'
+    # https://spec.commonmark.org/0.30/#atx-headings
+    # Note:
+    #   Consume spaces and tabs after opening hashes so entity.name
+    #   starts with first non-whitespace character,
+    #   but don't do so if directly followed by closing hashes
+    #   as terminator pattern requires them to match then.
+    - match: '[ \t]{,3}(#{1})(?:(?=[ \t]+#+[ \t]*$)|[ \t]+|$)'
       captures:
         1: punctuation.definition.heading.begin.markdown
-      push:
-        - meta_scope: markup.heading.1.markdown
-        - meta_content_scope: entity.name.section.markdown
-        - include: atx-heading-terminator
-        - include: inline-bold-italic
-    - match: '(#{2})(?!#)\s*(?=\S)'
+      push: atx-heading1-content
+    - match: '[ \t]{,3}(#{2})(?:(?=[ \t]+#+[ \t]*$)|[ \t]+|$)'
       captures:
         1: punctuation.definition.heading.begin.markdown
-      push:
-        - meta_scope: markup.heading.2.markdown
-        - meta_content_scope: entity.name.section.markdown
-        - include: atx-heading-terminator
-        - include: inline-bold-italic
-    - match: '(#{3})(?!#)\s*(?=\S)'
+      push: atx-heading2-content
+    - match: '[ \t]{,3}(#{3})(?:(?=[ \t]+#+[ \t]*$)|[ \t]+|$)'
       captures:
         1: punctuation.definition.heading.begin.markdown
-      push:
-        - meta_scope: markup.heading.3.markdown
-        - meta_content_scope: entity.name.section.markdown
-        - include: atx-heading-terminator
-        - include: inline-bold-italic
-    - match: '(#{4})(?!#)\s*(?=\S)'
+      push: atx-heading3-content
+    - match: '[ \t]{,3}(#{4})(?:(?=[ \t]+#+[ \t]*$)|[ \t]+|$)'
       captures:
         1: punctuation.definition.heading.begin.markdown
-      push:
-        - meta_scope: markup.heading.4.markdown
-        - meta_content_scope: entity.name.section.markdown
-        - include: atx-heading-terminator
-        - include: inline-bold-italic
-    - match: '(#{5})(?!#)\s*(?=\S)'
+      push: atx-heading4-content
+    - match: '[ \t]{,3}(#{5})(?:(?=[ \t]+#+[ \t]*$)|[ \t]+|$)'
       captures:
         1: punctuation.definition.heading.begin.markdown
-      push:
-        - meta_scope: markup.heading.5.markdown
-        - meta_content_scope: entity.name.section.markdown
-        - include: atx-heading-terminator
-        - include: inline-bold-italic
-    - match: '(#{6})(?!#)\s*(?=\S)'
+      push: atx-heading5-content
+    - match: '[ \t]{,3}(#{6})(?:(?=[ \t]+#+[ \t]*$)|[ \t]+|$)'
```

Comment:
maybe it makes sense to use a variable for the lookahead/end part as it is repeated 6 times in different match patterns